wandb/stats.py

```diff
@@ -40,4 +45,65 @@ def summary(self):
             'total_files': len(self._files),
             'uploaded_bytes': sum(f.uploaded for f in self._files.values()),
             'total_bytes': sum(f.size for f in self._files.values())
-        }
\ No newline at end of file
+        }
+
+
+class SystemStats(object):
+    def __init__(self, run):
+        try:
+            self.gpu_count = nvmlDeviceGetCount()
+        except NVMLError as err:
+            self.gpu_count = 0
+        self.run = run
+        self.sampler = {}
+        self.samples = 0
+        self._thread = threading.Thread(target=self._thread_body)
+        self._thread.daemon = True
+        self._thread.start()
+
+    def _thread_body(self):
+        while True:
+            stats = self.stats()
+            for stat, value in stats.items():
+                if isinstance(value, Number):
+                    self.sampler[stat] = self.sampler.get(stat, [])
+                    self.sampler[stat].append(value)
+            self.samples += 1
+            if self.samples >= 15:
+                self.flush()
+            time.sleep(2)
+
+    def flush(self):
+        stats = self.stats()
+        for stat, value in stats.items():
+            # TODO: a bit hacky, we assume all numbers should be averaged.  If you want
+            # max for a stat, you must put it in a sub key, like ["network"]["sent"]
+            if isinstance(value, Number):
+                samples = self.sampler.get(stat, [stats[stat]])
+                stats[stat] = round(reduce(lambda x, y: x + y,
```

Comment:
There is a race condition between this code and `.flush()`. The easiest way to fix it is probably to make a copy of `samples`, like `samples = list(samples)`.

Also btw you can do `sum(samples)` instead of using `reduce`.